### ISSUE TITLE
Logging

### ISSUE THREAD

**Issue by studywolf** [ctn-archive/nengo_theano#1]
Need to add logging functionality.

Doesn't seem like it will be too bad, but just need to make sure there's no confusion with the dt of the ensembles vs the dt of the network.run function. Should log from network.run. 

Do we want to take this implementation opportunity to change the way the spike logger logs spikes? Storing the index of the neurons that spike at each timestep rather than a giant matrix of zeros and ones?

**Comment by tbekolay** [comment]
Yeah, it would be good to talk about how we want to structure log files. I think that we want at least 3 ways of logging data:
1. To memory, for small networks and especially for simple unit testing.
2. To HDF5 files, for large networks with lots of data so that we can do lazy loading and other nice things. I strongly recommend that we use the [NeoHDF5](http://pythonhosted.org/neo/io.html#neo.io.NeoHdf5IO) format, which (I believe) is what PyNN uses as well, and will enable us to share data very easily with other Python neuroscience packages.
3. To CSV files, for a portable no-dependency solution that can be loaded anywhere (e.g., in Matlab).

Now, what those logs look like does depend a bit on the type of log file.

CSV files are necessarily row-oriented, and most naturally those rows represent one timestep. They're huge, but we can greatly increase the speed at which people can go through CSVs by making the length of each line predictable (so they can seek the file instead of reading it). Unfortunately, recording spikes makes that almost impossible. One workaround is to print the length of the line as the first column in the CSV file, meaning that people can read a tiny amount of data and still reliably skip all over the file.

Memory and HDF5 files need not be row-oriented, so they can be a bit smarter about how they store information.

**Comment by studywolf** [comment]
Also sidenote: Eric wrote a baller probe class that logs data, which was the part of the logger that I was looking forward to writing the least

**Comment by tbekolay** [comment]
For completeness, as discussed in issue #3 a possible fourth logging output would be a socket. There should also be a way to communicate the other way (provide input as well as gather output) over sockets, or perhaps shared memory.

**Comment by studywolf** [comment]
OK, I'm going to start on this. PyNN has it's own IO module that's different than the HDF5, but once it's loaded into Neo format it should be easy enough to just output it with the PyNN module instead. Seems like the main thing is going to be reading in the data to the appropriate Neo data structure, once that's done it should be straightforward.

**Comment by tbekolay** [comment]
Neo data structures are numpy subclasses, and are generally pretty easy to work with (though it might warrant some discussion as to the best way to organize our data using their data structures). But I made a loader based on the log file format that might be useful as an example of how to make these data structures. http://dl.dropbox.com/u/6182929/sim_io.py

The hard thing is socket IO, I think, but perhaps that should be a separate issue from this one.

**Comment by studywolf** [comment]
OK, after looking at this a bit, it seems like the thing to do is inside the net.make_probe function, depending on the data_type being recorded (at this point just a decoded or spikes signal) a Neo object is created to store the information (an AnalogSignal or SpikeTrain, respectively), and pass that to the probe for logging the data,  which I believe won't change the way any of the actual data is being stored in the probe, except that Neo seems to store the first input as the last entry in the array. 

Then, just for a first pass at this, I'll make another network function write_data or somesuch, and then the appropriate Neo Block - Segment - RecordingChannel / AnalogSignalArrays hierarchy is generated based on the set of probes created. Then you call this function after you've run the network and blamo, data saved to file in a nice format. 

From there then we can look at saving the data as the network is running, which will probably be able to work in with the socket logging code.

So this is my suggestion of a reasonable first pass at logging using the Neo structure. Thoughts? We'll more than likely rework the API entirely during the codesprint, but I just want to have something working as a base to start from. I'll start a new branch for this and start coding it up, but it would be cool to discuss better implementations.

**Comment by tbekolay** [comment]
This sounds like a good idea. My only hesitation is that it can be very expensive to append data to numpy arrays -- and it's very likely that the overhead of Neo (which tracks units for each array) might make appending even more expensive. Are we just using Python lists right now? If so, it might be quicker to keep them as lists and then in the `network.write_data` function make new Neo objects. Though, that'd require a data copy. Might be worth doing some benchmarking, but I know in the past I've sped up code a lot by not using numpy arrays when they change a lot.

**Comment by studywolf** [comment]
Right now the probe makes a numpy array size 1000 and either doubles or adds another 1000 whenever the limit is hit, I think the same thing can be done with the Neo objects, if I can then I'll do a first pass just with this and add a TODO for benchmarking. Also there must be something in Neo for handling appending efficiently, I'll look into it.

**Comment by tbekolay** [comment]
Ah okay, sounds good. I wouldn't count on there being anything in Neo to handling appending efficiently, it's not really designed for streaming, it's very much a write / read once type thing. It's a pretty small project as well, so I'd look more into NumPy itself for how to do it. Doubling the size when it hits a limit sounds like a pretty good strategy though!

**Comment by studywolf** [comment]
Actually thinking about it a bit more, having the probes keep lists is probably the better way to go. If they handle the appending well that's ideal. Then we can have a number of different write out functions if we so desire without wedding to any one in particular, and then Neo isn't an explicit dependency unless you want to write out. You would still be able to probe and plot without it. It seems like a more modular approach. And the write hdf5 function can just create the appropriate Neo objects from the probe data when it's called.

**Comment by tbekolay** [comment]
Right, that's a good point, a CSV file writer shouldn't need any dependencies.

**Comment by studywolf** [comment]
The Neo structures require a Quantities package labelling, any thought on what I should be using for spike trains and decoded signals? maybe uncertain or dimensionless for the decoded?

**Comment by tbekolay** [comment]
In the past, we've used seconds for spike trains. Decoded signals are dimensionless, as they don't really relate to any physical quatntites. But, it would be nice if it were possible to provide membrane potentials as mV.

**Comment by tbekolay** [comment]
Uncertain, I think, refers to quantities that have some degree of uncertainty; e.g., it's 6 +/- 0.1 degrees Celsius.

**Comment by studywolf** [comment]
alrighty! seconds, dimensionless, and mV it is!

**Comment by studywolf** [comment]
Basic logging writeout ability has been added! All probes data is written out to hdf5 file with `net.write_data_to_hdf5`
I'm not crazy about the name.
